lib/Index/IndexRecord.cpp

```diff
@@ -393,6 +396,195 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
+/// index store directory.
+///
+/// The swift interface file is emitted only when it doesn't exist yet, or when
+/// the PCM for the Clang module has been updated.
+///
+/// System modules without the 'cplusplus' requirement are not emitted.
+static void emitSymbolicInterfaceForClangModule(
+    ClangModuleUnit *clangModUnit, ModuleDecl *M,
+    const clang::Module *clangModule, StringRef indexStorePath,
+    const clang::CompilerInstance &clangCI, DiagnosticEngine &diags) {
+  if (!M->getASTContext().LangOpts.EnableCXXInterop)
+    return;
+  // Skip system modules without an explicit 'cplusplus' requirement.
+  bool isSystem = clangModUnit->isSystemModule();
+  if (isSystem && !importer::requiresCPlusPlus(clangModule))
+    return;
+
+  // Make sure the `interfaces` directory is created.
+  std::string error;
+  if (initSymbolicInterfaceStorePath(indexStorePath, error)) {
+    diags.diagnose(SourceLoc(), diag::error_create_index_dir, error);
+    return;
+  }
+
+  // Determine the output name for the symbolic interface file.
+  clang::serialization::ModuleFile *ModFile =
+      clangCI.getASTReader()->getModuleManager().lookup(
+          clangModule->getASTFile());
+  assert(ModFile && "no module file loaded for module ?");
+  SmallString<128> interfaceOutputPath = indexStorePath;
+  appendSymbolicInterfaceToIndexStorePath(interfaceOutputPath);
+  appendSymbolicInterfaceClangModuleFilename(ModFile->FileName,
+                                             interfaceOutputPath);
+
+  // Check if the symbolic interface file is already up to date.
+  auto upToDate = isFileUpToDateForOutputFile(
+      interfaceOutputPath, StringRef(ModFile->FileName), error);
+  if (!upToDate) {
+    diags.diagnose(SourceLoc(), diag::error_index_failed_status_check, error);
+    return;
+  }
+  if (M->getASTContext().LangOpts.EnableIndexingSystemModuleRemarks) {
+    diags.diagnose(SourceLoc(), diag::remark_emitting_symbolic_interface_module,
+                   interfaceOutputPath, *upToDate);
+  }
+  if (*upToDate)
+    return;
+
+  // Output the interface to a temporary file first.
+  SmallString<128> tempOutputPath;
+  tempOutputPath = llvm::sys::path::parent_path(interfaceOutputPath);
+  llvm::sys::path::append(tempOutputPath,
+                          llvm::sys::path::filename(interfaceOutputPath));
+  tempOutputPath += "-%%%%%%%%";
+  int tempFD;
+  if (llvm::sys::fs::createUniqueFile(tempOutputPath.str(), tempFD,
+                                      tempOutputPath)) {
+    llvm::raw_string_ostream errOS(error);
+    errOS << "failed to create temporary file: " << tempOutputPath;
+    diags.diagnose(SourceLoc(), diag::error_write_index_record, errOS.str());
+    return;
+  }
+
+  llvm::raw_fd_ostream os(tempFD, /*shouldClose=*/true);
+  std::unique_ptr<ASTPrinter> printer;
+  printer.reset(new StreamPrinter(os));
+  ide::printSymbolicSwiftClangModuleInterface(M, *printer, clangModule);
+  os.close();
+
```

Comment:
Should we make sure to delete the temporary file if anything up to and including the rename fails?

Reply:
good idea.